actors/migration/nv15/miner.go

```diff
@@ -19,63 +25,184 @@ type minerMigrator struct{}
+	outState.Deadlines = deadlinesOut
+
 	newHead, err := store.Put(ctx, &outState)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to flush outState: %w", err)
+	}
+
 	return &actorMigrationResult{
 		newCodeCID: m.migratedCodeCID(),
 		newHead:    newHead,
-	}, err
+	}, nil
 }
 
 func (m minerMigrator) migratedCodeCID() cid.Cid {
 	return builtin7.StorageMinerActorCodeID
 }
 
-func migrateSectors(ctx context.Context, store cbor.IpldStore, inRoot cid.Cid) (cid.Cid, error) {
-	ctxStore := adt.WrapStore(ctx, store)
-	inArray, err := adt.AsArray(ctxStore, inRoot, miner6.SectorsAmtBitwidth)
+func migrateSectors(ctx context.Context, store adt.Store, cache MigrationCache, minerAddr address.Address, inRoot cid.Cid) (cid.Cid, error) {
+	return cache.Load(CidKey(inRoot), func() (cid.Cid, error) {
+		inArray, err := adt.AsArray(store, inRoot, miner6.SectorsAmtBitwidth)
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("failed to read sectors array: %w", err)
+		}
+
+		okIn, prevInRoot, err := cache.Read(MinerPrevSectorsInKey(minerAddr))
+		if err != nil {
+			// TODO: Should we instead absorb errors like these and proceed cache-lessly?
+			return cid.Undef, xerrors.Errorf("failed to get previous inRoot from cache: %w", err)
+		}
+
+		okOut, prevOutRoot, err := cache.Read(MinerPrevSectorsOutKey(minerAddr))
+		if err != nil {
+			return cid.Undef, xerrors.Errorf("failed to get previous outRoot from cache: %w", err)
+		}
+
+		if okIn != okOut {
+			return cid.Undef, xerrors.Errorf("cannot have only one of the inRoot and outRoot in the cache")
+		}
+
+		var outArray *adt.Array
+		// first time we're doing this, do all the work
+		if !okIn {
+			outArray, err = adt.MakeEmptyArray(store, miner7.SectorsAmtBitwidth)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to construct new sectors array: %w", err)
+			}
+
+			var sectorInfo miner6.SectorOnChainInfo
+			if err = inArray.ForEach(&sectorInfo, func(k int64) error {
+				return outArray.Set(uint64(k), migrateSectorInfo(sectorInfo))
+			}); err != nil {
+				return cid.Undef, err
+			}
+		} else {
+
+			// we have previous work, but the AMT has changed -- diff them
+			diffs, err := amt.Diff(ctx, store, store, prevInRoot, inRoot, amt.UseTreeBitWidth(miner7.SectorsAmtBitwidth))
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to diff old and new Sector AMTs: %w", err)
+			}
+
+			inSectors, err := miner6.LoadSectors(store, inRoot)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to load inSectors: %w", err)
+			}
+
+			prevOutSectors, err := miner7.LoadSectors(store, prevOutRoot)
+			if err != nil {
+				return cid.Undef, xerrors.Errorf("failed to load prevOutSectors: %w", err)
+			}
+
+			for _, change := range diffs {
+				switch change.Type {
+				case amt.Remove:
+					if err := prevOutSectors.Delete(change.Key); err != nil {
+						return cid.Undef, xerrors.Errorf("failed to delete sector from prevOutSectors: %w", err)
+					}
+				case amt.Add:
```

Comment:
I think this needs to `fallthrough` otherwise you don't migrate sectors added to the amt in between.  https://go.dev/play/p/fPxSEXpevf7. This highlights the need for thorough testing.  A good place to draw inspiration is the deal label migration test: https://github.com/filecoin-project/specs-actors/pull/1551/files#diff-5f4089c6e3f8495e10bfa13e18ff8f0e34cb787c56256e0bfabc576f0253f66aR36